cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,195 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/status.h"
+#include "arrow/util/compression.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+namespace io {
+
+class InputStream;
+class OutputStream;
+class RandomAccessFile;
+
+}  // namespace io
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// nanoseconds since the epoch.
+using TimePoint =
+    std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+/// \brief EXPERIMENTAL: FileSystem entry type
+enum class ARROW_EXPORT FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+ARROW_EXPORT std::string ToString(FileType);
+
+/// \brief EXPERIMENTAL: FileSystem entry stats
+struct ARROW_EXPORT FileStats {
+  FileStats() = default;
+  FileStats(FileStats&&) = default;
+  FileStats& operator=(FileStats&&) = default;
+  FileStats(const FileStats&) = default;
+  FileStats& operator=(const FileStats&) = default;
+
+  // The file type.
+  FileType type() const { return type_; }
+  void type(FileType type) { type_ = type; }
+
+  // The full file path in the filesystem.
+  std::string path() const { return path_; }
+  void path(const std::string& path) { path_ = path; }
+
+  // The file base name (component after the last directory separator or '/').
+  std::string base_name() const;
+
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size() const { return size_; }
+  void size(int64_t size) { size_ = size; }
+
+  // The time of last modification, if available.
+  TimePoint mtime() const { return mtime_; }
+  void mtime(TimePoint mtime) { mtime_ = mtime; }
+
+ protected:
+  FileType type_ = FileType::Unknown;
+  std::string path_;
+  int64_t size_ = -1;
+  TimePoint mtime_ = TimePoint(TimePoint::duration(-1));
+};
+
+/// \brief EXPERIMENTAL: stream / file open flags
+struct ARROW_EXPORT OpenFlags {
+  // Whether to transparently compress / decompress.
+  Compression::type compression = Compression::DETECT;
+
+  OpenFlags() {}
+};
+
+/// \brief EXPERIMENTAL: file selector
+struct ARROW_EXPORT Selector {
+  // The directory in which to select files.
+  // If the path exists but doesn't point to a directory, this should be an error.
+  std::string base_dir;
```

Comment:
Are you thinking about some kind of wildcard API at some point?

Reply:
Yes, that's the idea, though it depends whether it's useful to have filesystem-specific implementations.